url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/eventlog/CorrectionGenerator.java

```diff
@@ -0,0 +1,400 @@
+				logger.info("Generated strict correction from difference " + diff.getID() + ":\n\t" + corr);
+				
+				Correction identical = null;
+				for (Correction existingCorr : corrections) {
+					if (existingCorr.equals(corr)) {
+						identical = existingCorr;
+						break;
+					}
+				}
+				if (identical != null) {
+					logger.info("Correction " + corr.getName() 
+							+ " is identical to previously generated correction " 
+							+ identical.getName() + ", skipping " + corr.getName());
+				} else {
+					percentages.put(corr, relevantDiffs.get(diff));
+					corrections.add(corr);
+				}
+			}
+		}
+		
+		List<Correction> updatedCorrections = this.removeOverlaps(corrections, percentages);
+		for (Correction c : updatedCorrections) {
+			logger.info("Final " + c);
+		}
+		return updatedCorrections;
+	}
+	
+	
+	public List<TraceDifference> getDifferencesWithTraces(Map<TraceDifference, Double> diffsWithOccurence) {
+		this.setStrictMode(true);
+		
+		diffsWithOccurence.putAll(diffExplorer.computeRelevantDifferences());
+		
+		List<TraceDifference> orderedDiffs = diffExplorer.orderByExecutedTraceSize(
+				new ArrayList<TraceDifference>(diffsWithOccurence.keySet()));
+		return orderedDiffs;
+	}
+	
+	public List<TraceDifference> getDifferencesWithoutTraces(Map<TraceDifference, Double> diffsWithOccurence) {
+		this.setStrictMode(false);
+		
+		diffsWithOccurence.putAll(diffExplorer.computeRelevantDifferences());
+		
+		List<TraceDifference> orderedDiffs = diffExplorer.orderByExecutedTraceSize(
+				new ArrayList<TraceDifference>(diffsWithOccurence.keySet()));
+		return orderedDiffs;
+	}
+	
+	
+	public List<Correction> generateRelevantRelaxedCorrections(Map<TraceDifference, Double> relevantDiffs, 
+			List<TraceDifference> orderedDiffs) throws CorrectionGenerationException {
+		this.setStrictMode(false);
+		List<Correction> corrections = new ArrayList<Correction>();
+		
+		Map<Correction, Double> percentages = new HashMap<Correction, Double>();
+		for (TraceDifference diff : orderedDiffs) {	
+			Map<Correction, Correction> matches = new HashMap<Correction, Correction>();
+			List<Correction> newCorrections = diff.generateCorrections(Type.RELAXED, model, allActivities);
+			
+			for (Correction corr : newCorrections) {
+				logger.info("Generated relaxed correction from difference " + diff.getID() + ":\n\t" + corr);
+				
+				Correction identical = null;
+				for (Correction existingCorr : corrections) {
+					if (existingCorr.equals(corr)) {
+						identical = existingCorr;
+						matches.put(corr, existingCorr);
+						break;
+					}
+					if (corr.isSameExceptForFromNode(existingCorr)) {
+						ActivityNode fromNode1 = corr.getAdaptation().getFromNode();
+						ActivityNode fromNode2 = existingCorr.getAdaptation().getFromNode();
+						
+						for (Correction c1 : matches.keySet()) {
+							if (c1.getAdaptation().getAdaptationModel().containsProcessNode(fromNode1)) {
+								Correction c2 = matches.get(c1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.